### PR TITLE
Wireup the dqlite minimal app and unit services

### PR DESCRIPTION
**Pull request by wallyworld**
This PR wires up the dqlite backed application and unit services to add and delete apps and units in dqlite to correspond with those in mongo state. The integration points are:
- bootstrap (add controller app and unit)
- deploy (add application api)
- add unit
- cleanup (remove app and unit)
- enable ha (add new controller units)
- migration model import

There's a little bit of churn in tests because of the helper AddApplication method now needing to take a mock for the dqlite service (since the real code dual writes the application data). This will all go away as we move away from mongo. What's here is just the minimum to integrate with the code which updates mongo state since it's all being rewritten.

During QA, an issue was discovered with migration - the service factory passed to the importer had been set up for the controller model, not the target model being migrated. This meant that the dqlite records for the machines, app, units were inserted into the controller model. So the apiserver root needs to be able to hand out a service factory getter to enable the model imported to update the correct dqlite database. This change is in the [second commit](https://github.com/juju/juju/pull/16887/commits/0b172b3896c49941012eb20e63c54df883e25019).

## QA steps

bootstrap
add model
deploy ubuntu -n 2
add-unit ubuntu
```
dqlite> select * from application;
53533bd4-038c-4148-8add-1d191a5007c2|ubuntu|0
dqlite> select * from unit;
c7835d21-2aed-4761-8c6f-20dfbc47f89d|ubuntu/0|53533bd4-038c-4148-8add-1d191a5007c2|eed19bd0-968d-453b-802d-d3a67fecceb7|0
4b975791-d3d2-4408-87f7-98311e6dbd75|ubuntu/1|53533bd4-038c-4148-8add-1d191a5007c2|24c4f0d4-fbf3-4717-875a-f607a04e0196|0
88054032-2658-4177-8a19-53740560eff8|ubuntu/2|53533bd4-038c-4148-8add-1d191a5007c2|6dc9175d-89ae-4e2b-87a9-222cfdf6050b|0
```

enable-ha
```
dqlite> select * from unit;
ac840fda-7ce0-43d0-8a2c-689868ab133e|controller/0|143992fc-232f-49e0-8904-d48de436983c|1b53a18a-a51f-4a21-88c8-88e98896f2c7|0
52a4a0ad-b1b2-424c-88bb-b09256e70cf2|controller/1|143992fc-232f-49e0-8904-d48de436983c|15d245a0-e7b4-4ded-88c5-936213464a7a|0
ef66de60-8509-4605-8cc5-6158c6a77241|controller/2|143992fc-232f-49e0-8904-d48de436983c|41bfa08c-f636-4dec-8ea9-ed5104daaed0|0
```

migrate model, check app and unit tables in the target model

delete unit and application and check rows are removed from dqlite

## Links

**Jira card:** JUJU-5406

### CODE REVIEW

File: apiserver/root.go [hunk 1]
```diff
@@ -828,9 +840,10 @@ func (ctx *facadeContext) ModelImporter() facade.ModelImporter {
 	pool := ctx.r.shared.statePool
 	return migration.NewModelImporter(
 		state.NewController(pool),
-		ctx.migrationScope(),
+		ctx.migrationScope,
+		ctx.ServiceFactory().ModelManager(),
 		ctx.ServiceFactory().ControllerConfig(),
-		ctx.ServiceFactory().Machine(),
+		ctx.ServiceFactoryGetter(),
```
Comment (SimonRichardson) [inline apiserver/root.go:846]: Is there any way to not expose this as a public method? I've nicely hidden this behind a public API `ModelExporter` and `ModelImporter` on the `facade.Context` interface.

I believe you could even do `ctx.r.serviceFactoryGetter` and you get the win win.

**Review by SimonRichardson — CHANGES_REQUESTED** [review]
I don't want to expose `ServiceFactoryGetter` on the `facadea.Context` interface as it allows facades to access models which they don't own. I understand that model migration needs this information. So having it nicely encapsulated behind `ModelImporter` is sufficient here.

File: apiserver/facade/interface.go [hunk 2]
```diff
@@ -194,6 +194,8 @@ type ModelMigrationFactory interface {
 type ServiceFactory interface {
 	// ServiceFactory returns the services factory for the current model.
 	ServiceFactory() servicefactory.ServiceFactory
+
+	ServiceFactoryGetter() servicefactory.ServiceFactoryGetter
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facade/interface.go:198]: Controller facades will need something like this. Controller facades do need to access arbitrary models. So we can remove this now but we'll need a solution later when we implement cross model relations facades for example. Maybe we can find a way to only expose this to controller facades.

[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facade/interface.go:198]: Perfect. Let's work on that when it comes up. 👍

File: apiserver/root.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/root.go]: done

File: apiserver/facades/client/application/deploy.go [hunk 3]
```diff
@@ -70,10 +71,16 @@ type MachineSaver interface {
 	Save(context.Context, string) error
 }
 
+// ApplicationSaver instances save an application to dqlite state.
+type ApplicationSaver interface {
+	Save(ctx context.Context, name string, units ...applicationservice.AddUnitParams) error
```
Comment (SimonRichardson) [inline apiserver/facades/client/application/deploy.go:76]: It's been noted that these should be `SaveApplication`. We're getting rid of just verbs for methods.

File: apiserver/apiserver.go [hunk 4]
```diff
@@ -1158,7 +1158,7 @@ func (srv *Server) serveConn(
 	st, err := statePool.Get(resolvedModelUUID)
 	if err == nil {
 		defer st.Release()
-		handler, err = newAPIHandler(srv, st.State, conn, serviceFactory, tracer, objectStore, controllerObjectStore, modelUUID, connectionID, host)
+		handler, err = newAPIHandler(srv, st.State, conn, serviceFactory, srv.shared.serviceFactoryGetter, tracer, objectStore, controllerObjectStore, modelUUID, connectionID, host)
```
Comment (SimonRichardson) [inline apiserver/apiserver.go:1161]: Thanks this is much better.

File: apiserver/common/remove.go [hunk 5]
```diff
@@ -17,30 +17,38 @@ import (
 	"github.com/juju/juju/state"
 )
 
+// UnitRemover deletes a unit from the dqlite database.
+// This allows us to initially weave some dqlite support into the cleanup workflow.
+type UnitRemover interface {
+	Delete(context.Context, string) error
```
Comment (SimonRichardson) [inline apiserver/common/remove.go:23]: This should be `DeleteUnit`

**Review by SimonRichardson — APPROVED** [review]
I'm fine with the code changes in this PR, but we need to sweep back around and update the one word verb method names from the services. @manadart has stated that the it should be verb then entity; `SaveMachine` and `DeleteUnit`.

**Pull request merged**